backend/icpy/services/chat_service.py

```
@@ -716,32 +737,96 @@ async def _process_with_custom_agent(self, user_message: ChatMessage, agent_type
+                                    logger.info(f"Custom agent: Attempting to embed explorer image from absolute path: {absolute_field}")
+                                    import os
+                                    from pathlib import Path
+                                    ws_root = os.environ.get('WORKSPACE_ROOT')
+                                    abs_path = Path(absolute_field).resolve()
+                                    # Only allow reading from within WORKSPACE_ROOT when set
+                                    if ws_root:
+                                        ws_root_p = Path(ws_root).resolve()
+                                        abs_path.relative_to(ws_root_p)
+                                    # If WORKSPACE_ROOT is not set, as a safety fallback disallow absolute paths outside cwd/workspace
+                                    elif not abs_path.is_file():
+                                        raise ValueError('Invalid path')
```

Ruff (0.13.1)

Abstract `raise` to an inner function

(TRY301)